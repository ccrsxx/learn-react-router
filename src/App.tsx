import { useState } from 'react';
import {
  Routes,
  Route,
  useSearchParams,
  useNavigate,
  useLocation
} from 'react-router-dom';
import { invoices as invoicesData, getData, deleteData } from './common';
import { Header, Footer, EmptyInvoice, EmptyPage } from './components';
import { Homepage, Invoices, Invoice, Expenses } from './routes';

export function App() {
  const [count, setCount] = useState(0);
  const [invoices, setInvoices] = useState(invoicesData);
  const [searchParams, setSearchParams] = useSearchParams();

  const [navigate, { search }] = [useNavigate(), useLocation()];

  const incrementCount = () => setCount(count + 1);
  const decrementCount = () => setCount(count - 1);

  const handleParamsChange = ({
    target: { value }
  }: React.ChangeEvent<HTMLInputElement>) =>
    value ? setSearchParams({ filter: value }) : setSearchParams({});

  const getInvoice = (id: number) => getData(invoices, id);

  const deleteInvoice = (id: number) => () => {
    setInvoices(deleteData(invoices, id));
    navigate(`/invoices${search}`);
  };

  const paramsValue = searchParams.get('filter') ?? '';

  return (
    <div className='flex min-h-screen flex-col items-center gap-6 py-8'>
      <Header />
      <Routes>
        <Route path='/' element={<Homepage />} />
        <Route
          path='expenses'
          element={
            <Expenses
              count={count}
              incrementCount={incrementCount}
              decrementCount={decrementCount}
            />
          }
        />
        <Route
          path='invoices'
          element={
            <Invoices
              invoices={invoices}
              paramsValue={paramsValue}
              handleParamsChange={handleParamsChange}
            />
          }
        >
          <Route index element={<EmptyInvoice />} />
          <Route
            path=':invoiceId'
            element={
              <Invoice getInvoice={getInvoice} deleteInvoice={deleteInvoice} />
            }
          />
        </Route>
        <Route path='*' element={<EmptyPage />} />
      </Routes>
      <Footer />
    </div>
  );
}
