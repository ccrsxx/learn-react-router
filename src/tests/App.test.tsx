import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { App } from '../App';

describe('test App component', () => {
  beforeEach(() => {
    render(<App />, { wrapper: Router });
  });

  it('matches previous snapshot', () => {
    const container = document.querySelector('body > div > div');
    expect(container).toMatchSnapshot();
  });

  it('can go to multiple routes', () => {
    const homepageButton = screen.getByRole('link', { name: /app/i });
    const expensesButton = screen.getByRole('link', { name: /expenses/i });
    const invoicesButton = screen.getByRole('link', { name: /invoices/i });

    userEvent.click(homepageButton);

    expect(
      screen.getByRole('heading', { name: /welcome to react/i })
    ).toBeInTheDocument();

    userEvent.click(expensesButton);

    expect(screen.getByText(/count/i)).toBeInTheDocument();

    userEvent.click(invoicesButton);

    expect(screen.getByText(/ocean/i)).toBeInTheDocument();
  });

  it('can increment and decrement states on expenses route', () => {
    function getSpanCount() {
      return screen.getByText(
        (content, element) =>
          element?.tagName === 'SPAN' && /^$|\d+/.test(content)
      ).textContent;
    }

    function clickCount(times: number, button: HTMLElement) {
      [...Array(times)].forEach(() => userEvent.click(button));
    }

    const expensesButton = screen.getByRole('link', { name: /expenses/i });

    userEvent.click(expensesButton);

    const incrementButton = screen.getByText(/increment/i);
    const decrementButton = screen.getByText(/decrement/i);

    clickCount(1, incrementButton);
    expect(getSpanCount()).toBe('1');

    clickCount(1, decrementButton);
    expect(getSpanCount()).toBe('0');

    clickCount(10, incrementButton);
    expect(getSpanCount()).toBe('10');

    clickCount(40, decrementButton);
    expect(getSpanCount()).toBe('-30');
  });

  it('can filter invoices', () => {
    const invoicesButton = screen.getByRole('link', { name: /invoices/i });
    const invoiceBox = () => document.querySelector('div > h2') as HTMLElement;

    userEvent.click(invoicesButton);

    expect(invoiceBox().textContent).toBe('ðŸ¤¡ No selected invoice');

    const oceanInvoice = screen.getByRole('link', { name: /ocean/i });

    userEvent.click(oceanInvoice);

    expect(invoiceBox().textContent).toBe('Invoice#2003');
  });
});
