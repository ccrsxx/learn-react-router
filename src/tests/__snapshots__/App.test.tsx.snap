// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test App component matches previous snapshot 1`] = `
<div
  class="flex min-h-screen flex-col items-center gap-6 py-8"
>
  <header
    class="flex flex-col items-center gap-4"
  >
    <h1
      class="text-2xl font-bold"
    >
      Welcome to React Router!
    </h1>
    <nav
      class="flex w-full gap-4 border-b-2 pb-4 text-white"
    >
      <a
        aria-current="page"
        class="scale-110 font-bold text-black ring-2 ring-pink-400 ring-offset-2 w-full max-w-xs rounded-md bg-green-400 px-2 py-1 text-center transition"
        href="/"
      >
        App
      </a>
      <a
        class="false w-full max-w-xs rounded-md bg-green-400 px-2 py-1 text-center transition"
        href="/expenses"
      >
        Expenses
      </a>
      <a
        class="false w-full max-w-xs rounded-md bg-green-400 px-2 py-1 text-center transition"
        href="/invoices"
      >
        Invoices
      </a>
    </nav>
  </header>
  <main
    class="flex flex-1 animate-fade flex-col gap-4 text-center"
  >
    <h2
      class="border-b pb-1 text-lg font-medium"
    >
      Welcome to the my first React Router App!
    </h2>
    <p>
      this is from form / (root of the project)
    </p>
  </main>
  <footer
    class="text-lg"
  >
    <p>
      Made with
       
      <span
        aria-label="heart"
        role="img"
      >
        ❤️
         
      </span>
      by
       
      <a
        class="rounded px-1 pb-[2px] font-medium text-blue-400 decoration-transparent underline-offset-2 transition-all duration-300 hover:underline hover:decoration-pink-200 hover:brightness-125  focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-pink-400  focus-visible:brightness-125"
        href="https://github.com/ccrsxx"
      >
        ccrsxx
      </a>
    </p>
  </footer>
</div>
`;
